### PR TITLE
REALMC-6991 Add topology check to client

### PR DESCRIPTION
**Pull request by tsedgwick**

### CODE REVIEW

File: x/mongo/driver/topology/topology.go [hunk 1]
```diff
@@ -673,3 +673,41 @@ func (t *Topology) String() string {
...
+	}
+
+	t.serversLock.Lock()
+	serversCopy := make([]*Server, 0, len(t.servers))
+	for _, s := range t.servers {
+		serversCopy = append(serversCopy, s)
+	}
+	t.serversLock.Unlock()
+
+	if len(serversCopy) == 0 || len(serversCopy) == 1 {
```
Comment (mpobrien) [inline x/mongo/driver/topology/topology.go:694]: q: isn't it still "inconsistent" if `len(serversCopy) == 1` and the one server listed is in the "primary" state when the description says `ReplicaSetNoPrimary`?
or i guess to put it a different way, is there any specific reason why the criteria applied in the loop below can't also apply to a one-node topology?

**Review by mpobrien — APPROVED** [review]

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (tsedgwick) [reply x/mongo/driver/topology/topology.go:694]: that's true, but is it possible for a one-node topology to be listed as a secondary?

File: x/mongo/driver/topology/topology.go [hunk 2]
```diff
@@ -673,3 +673,41 @@ func (t *Topology) String() string {
...
+		return true
+	}
+	var foundPrimary bool
+	for _, server := range serversCopy {
+		serverDesc := server.Description()
+		switch serverDesc.Kind {
+		case description.RSPrimary:
+			if foundPrimary {
+				return true
```
Comment (mpobrien) [inline x/mongo/driver/topology/topology.go:703]: i'm confused by this part - it looks like this path would only ever be reached if there were *two* nodes in the list that have `RSPrimary`.

edit: Actually, shouldn't the body of this `case` statement just be `return false` ?
at this point, we know that the topology claims to have no primary, but there is one, so it's not consistent - no need to continue looping further

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (mpobrien) [reply x/mongo/driver/topology/topology.go:694]: it's very unlikely, but the node could also be in any other non-primary state (startup, recovering, down, etc.) - not just "secondary"

Comment (tsedgwick) [reply x/mongo/driver/topology/topology.go:694]: interesting, that makes sense. If the only node was in another state, I would expect the client to fail on connections anyhow - regardless of if we tried to reconnect using this consistency check.

**Pull request merged**